Provide a default implementation of multipliedFullWidth

stephentyrone opened the pull request:
Previously, [U]Int64 fatalErrored on 32b platforms, which is obviously undesirable. This PR provides a default implementation on FixedWidthInteger, which is not ideally efficient for all types, but is correct, and gives the optimizer all the information that it needs to generate good code in the important case of Int64 arithmetic on 32b platforms. There's still some minor room for improvement, but we'll call that an optimizer bug now.

Fixes #53300

stephentyrone committed:
Provide a default implementation of multipliedFullWidth

Previously, [U]Int64 fatalErrored on 32b platforms, which is obviously undesirable. This PR provides a default implementation on FixedWidthInteger, which is not ideally efficient for all types, but is correct, and gives the optimizer all the information that it needs to generate good code in the important case of Int64 arithmetic on 32b platforms. There's still some minor room for improvement, but we'll call that an optimizer bug now.

stephentyrone commented:
I still need to add some test cases for this. In the meantime, @palimondo, @Azoy, @xwu, @natecook1000, I would love feedback on any details that you'd like me to document more clearly.

Azoy commented on stdlib/public/core/Integers.swift:
```diff
@@ -2317,6 +2317,59 @@ extension FixedWidthInteger {
+  // own implementation, but neither will it be catastrophically bad. It can
+  // surely be improved on even for Int64, but that is mostly an optimization
+  // problem; the basic algorithm here gives the compiler all the information
+  // that it needs to generate efficient code.
```
The commentary is really nice, but shouldn't the documentation be placed above the function? I imagine you can put the commentary right above the docs.

stephentyrone replied:
The documentation is provided by the protocol. It doesn't need to be on every implementation.

Azoy replied:
Good to know!

gottesmm commented:
Can you add some sort of test to make sure that the optimizer does not regress?

stephentyrone commented:
> Can you add some sort of test to make sure that the optimizer does not regress?

I think we have pretty good coverage for the patterns in question in the LLVM test suite already, but I'll look into it.

xwu commented:
Both the code and the comments are very reasonable. For the fleeting moment that `DoubleWidth` was completely implemented in the standard library, I'd had this operation defined for 64-bit integers on 32-bit platforms in terms of `DoubleWidth`, but alas. Glad to see it's taken care of again.

stephentyrone committed:
Clarify comments somewhat, remove `merge` nested function

I was only using `merge` in one place, so making it a function seems unnecessary. Also got rid of some trucatingIfNeeded inits where the compiler is able to reason that no checks are needed anyway.

palimondo commented on stdlib/public/core/Integers.swift:
```diff
@@ -2328,23 +2328,25 @@ extension FixedWidthInteger {
-    func merge<T: FixedWidthInteger>(_ high: T, _ low: T.Magnitude) -> T {
-      let n = T.bitWidth/2
-      return high &<< n | T(truncatingIfNeeded: low)
+      return (x >> n, T.Magnitude(truncatingIfNeeded: x) & ((1 &<< n) &- 1))
```
Why is the truncation init necessary here, if it was not below? Isn't `T.Magnitude` "wider" than `T` already? 

<sub>⚠️ I'm talking from perspective of new learner here, I really dunno what I'm talking about…</sub>

stephentyrone replied:
It’s necessary here to handle negative values of ‘x’ correctly. That never happens in the other cases.

palimondo replied:
OK, reading the comment on fa3a3d8, `init:truncatingIfNeeded` does, even though it's more verbose, actually produce less code by omitting the checks. You went with `init` below for better code clarity (less verbose), when you were certain the optimized will strip the checks. 

Swift's safety/performance and code brevity/clarity tradeoffs are quite counter intuitive here. But this does kill perf of unoptimized builds, right?

palimondo replied:
Swift's taken the safety first mantra really seriously.

I mean, we use the more verbose API for less safe operations, but in the end, even when wielded by an expert like you, there is value in keeping the expressions mathematically concise, to give bugs less place to hide in cleaner code. 😢

It's as if the right way to code this was to express the idea first in most concise notation and *than* go in with an axe and butcher the clean math for performance's sake by adding all the unsafe `truncatungIfNeeded` garbage. But somehow it all feels wrong…

I'm always concerned with the performance impact on the unoptimized debug builds. I feel like we need some `@optimize(always)` attribute with some a compiler flag for really debugging this in development for such basic and performance critical building blocks as this. It always bothers me that `for-in` beats functional code with `map`/`reduce` just because they perform abysmally in unoptimized builds. 😤

stephentyrone replied:
The perf for unoptimized builds can't be killed: it's already dead because inlining and hence specialization isn't going to happen for this function. That's OK though, it's still better than a fatalError (and short of adding a new @semantics annotation, we can't do anything about it yet).

stephentyrone replied:
> It's as if the right way to code this was to express the idea first in most concise notation and then go in with an axe and butcher the clean math for performance's sake by adding all the unsafe truncatingIfNeeded garbage.

In the long-term, the right way to address this is probably to define some stdlib-internal integer types that always wrapping arithmetic. That's a useful building block to have for library development, and gets us to a best-of-both worlds situation. In the shorter term, keep in mind that `truncatingIfNeeded` and wrapping arithmetic are **not** just "turning off safety for performance reasons". They are semantically different operations, still with fully defined behavior; they are perfectly safe, and just mean something else.

stephentyrone committed:
Add some basic test coverage specifically for multipliedFullWidth

stephentyrone commented:
@swift-ci please smoke test and merge

palimondo commented on test/stdlib/MultipliedFullWidth.swift:
```diff
@@ -0,0 +1,127 @@
+//===--- MultiplieddFullWidth.swift.gyb -----------------------*- swift -*-===//
```
`MultiplieddFullWidth` has a double `dd` typo.

stephentyrone committed:
Fix typo, further clarify bounds comments.

stephentyrone commented:
@swift-ci please smoke test and merge

stephentyrone commented:
@swift-ci please smoke test and merge

stephentyrone committed:
Make new defaulted implementation @_aEIC so we don't need availability.

stephentyrone commented:
@swift-ci please smoke test

stephentyrone commented:
@swift-ci smoke test macOS

natecook1000 reviewed (approved):
Looks great – glad to see this in!
